pyTigerGraph/schema.py

```diff
@@ -0,0 +1,268 @@
+
+BASE_TYPES  = ["string", "int", "uint", "float", "double", "bool", "datetime"]
+PRIMARY_ID_TYPES = ["string", "int", "uint", "datetime"]
+COLLECTION_TYPES = ["list", "set", "map"]
+COLLECTION_VALUE_TYPES = ["int", "double", "float", "string", "datetime", "udt"]
+MAP_KEY_TYPES = ["int", "string", "datetime"]
+
+def _parse_type(attr):
+    collection_types = ""
+    if attr["AttributeType"].get("ValueTypeName"):
+        if attr["AttributeType"].get("KeyTypeName"):
+            collection_types += "<"+ attr["AttributeType"].get("KeyTypeName") + "," + attr["AttributeType"].get("ValueTypeName") + ">"
+        else:
+            collection_types += "<"+attr["AttributeType"].get("ValueTypeName") + ">"
+    attr_type = (attr["AttributeType"]["Name"] + collection_types).upper()
+    return attr_type
+
+def _get_type(attr_type):
+    if attr_type == "STRING":
+        return str
+    elif attr_type == "INT":
+        return int
+    elif attr_type == "FLOAT":
+        return float
+    elif "LIST" in attr_type:
+        val_type = attr_type.split("<")[1].strip(">")
+        return List[_get_type(val_type)]
+    elif "MAP" in attr_type:
+        key_val = attr_type.split("<")[1].strip(">")
+        key_type = key_val.split(",")[0]
+        val_type = key_val.split(",")[1]
+        return Dict[_get_type(key_type), _get_type(val_type)]
+    elif attr_type == "BOOL":
+        return bool
+    elif attr_type == "datetime":
+        return datetime
+    else:
+        return attr_type
+
+
+def _py_to_tg_type(attr_type):
+    if attr_type == str:
+        return "STRING"
+    elif attr_type == int:
+        return "INT"
+    elif attr_type == float:
+        return "FLOAT"
+    elif attr_type == list:
+        raise TigerGraphException("Must define value type within list")
+    elif attr_type == dict:
+        raise TigerGraphException("Must define key and value types within dictionary/map")
+    elif attr_type == datetime:
+        return "DATETIME"
+    elif (str(type(attr_type)) == "<class 'typing._GenericAlias'>") and attr_type._name == "List":
+        val_type = _py_to_tg_type(attr_type.__args__[0])
+        if val_type.lower() in COLLECTION_VALUE_TYPES:
+            return "LIST<"+val_type+">"
+        else:
+            raise TigerGraphException(val_type + " not a valid type for the value type in LISTs.")
+    elif (str(type(attr_type)) == "<class 'typing._GenericAlias'>") and attr_type._name == "Dict":
+        key_type = _py_to_tg_type(attr_type.__args__[0])
+        val_type = _py_to_tg_type(attr_type.__args__[1])
+        if key_type.lower() in MAP_KEY_TYPES:
+            if val_type.lower() in COLLECTION_VALUE_TYPES:
+                return "MAP<"+key_type+","+val_type+">"
+            else:
+                raise TigerGraphException(val_type + " not a valid type for the value type in MAPs.")
+        else:
+            raise TigerGraphException(key_type + " not a valid type for the key type in MAPs.")
+    else:
+        print(attr_type)
```

Comment:
Is this intended for debugging or actually in prod?

Reply:
Whoops, yeah debug